OCULUS_multiview implementation w/ multisampling support

snagy opened the pull request:
**Related issue:** #20368

This is a crack at reimplementing multiview for XR devices that support the OCULUS_multiview extension.  This extension lets us use multiview along with MSAA.

This is a huge speed improvement for applications that are CPU limited and draw count bound.  Even the VR ballshooter example runs at a higher framerate when shooting on the quest 2 with multiview enabled than without.

Unfortunately the extension has one big problem - the multisampled render to texture extension this uses will discard the frame buffer if other texture operations are used during rendering.  This is outlined in https://github.com/KhronosGroup/WebGL/issues/2912 .  I've added a deferred upload of texture data (such as video frames) to mitigate this problem, by deferring uploads until after rendering of the main scene has finished.  This adds a frame of latency to the texture uploads, and is only active if multiview is enabled.

Some apps (mostly ones that render stuff like real time reflections to a texture) won't work with this extension without refactoring when those renders happen, to ensure that they don't interrupt the main frame.

Most experiences will get a free perf gain from this extension, but some will get broken.  Because of this, I do not enable multiview by default and instead added a 'multiviewStereo' flag to the renderer properties to enable it.  If this flag isn't set, everything should work as before.

Mugen87 commented on src/renderers/webgl/WebGLTextures.js:
```diff
@@ -1273,7 +1333,54 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
+
+				} else if ( depthTexture.type === UnsignedIntType ) {
+
+					glInternalFormat = _gl.DEPTH_COMPONENT24;
```
DeepScan complains here since `glInternalFormat` already has the value of `_gl.DEPTH_COMPONENT24`.

https://deepscan.io/dashboard/#view=project&pid=19901&iid=120730212&tid=16600&subview=pull-request&prid=1229149

Mugen87 commented:
It seems the PR modifies some example files which are unrelated to this change (like `BufferGeometryUtils`, `HTMLMesh` etc.). Even these modifications are correct, it still clearer to not include them in the PR.

LeviPesin commented:
Please also remove build and *.d.ts files.

mrdoob commented on examples/webxr_vr_ballshooter.html:
```diff
@@ -90,7 +90,7 @@
 				//
 
-				renderer = new THREE.WebGLRenderer( { antialias: true } );
+				renderer = new THREE.WebGLRenderer( { antialias: true, multiviewStereo: true } );
```
I don't think the developer should have to enable this.
Ideally the renderer would just use the extension when available.

mrdoob commented:
> Unfortunately the extension has one big problem - the multisampled render to texture extension this uses will discard the frame buffer if other texture operations are used during rendering.  This is outlined in https://github.com/KhronosGroup/WebGL/issues/2912 .  I've added a deferred upload of texture data (such as video frames) to mitigate this problem, by deferring uploads until after rendering of the main scene has finished.  This adds a frame of latency to the texture uploads, and is only active if multiview is enabled.

Do you know if this is something that will get fixed in the spec somehow?

Ideally this is something that the renderer would just use when available and the developer shouldn't have to learn all these details and have to choose between higher CPU usage or being a frame late...

snagy commented:
> Do you know if this is something that will get fixed in the spec somehow?
> 
> Ideally this is something that the renderer would just use when available and the developer shouldn't have to learn all these details and have to choose between higher CPU usage or being a frame late...

I agree completely that I want this to be automatic and error free.  From my understanding, fixing it at a spec level isn't trivial - you've basically got to pick your poison between slowdown and shimmering (with automatically handled resolves) or failed rendering (with discard on resolve).  AFAICT, both options require changes to how three handles texture uploads to behave optimally.

Because this is such a big win for experiences that can use it, I'm inclined to check it in behind a flag so folks can opt-in and see if the limitations are acceptable for their app.  I'll also keep looking at how we can make it easier to use, both on the platform side and three.js, and see if I can find a generalized solution for everyone that doesn't involve the upload delay.

mrdoob commented:
> From my understanding, fixing it at a spec level isn't trivial - you've basically got to pick your poison between slowdown and shimmering (with automatically handled resolves) or failed rendering (with discard on resolve).

@rcabanier Do you have any insights on this?

cabanier commented:
> > From my understanding, fixing it at a spec level isn't trivial - you've basically got to pick your poison between slowdown and shimmering (with automatically handled resolves) or failed rendering (with discard on resolve).
> 
> @rcabanier Do you have any insights on this?

It is unlikely that this will be fixed in the OpenGL spec. Just as with [ext_multisampled_render_to_texture](https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_multisampled_render_to_texture.txt) this was done on *purpose* because storing/loading and binding/unbinding are very costly on mobile hardware.
By putting in this flushing behavior, developers are forced to update their code so they don't run into slowdowns.

That being said, when we switch to the Vulkan backend, we will have more control over loading and storing so we can fix it ourselves.

cabanier commented:
Having multisampled multiview as an option will definitely be very helpful. It's hard to have complex immersive scenes on our limited hardware and this feature greatly helps code that is CPU limited.

mrdoob commented:
@cabanier Thanks!

What about this other issue?

> Unfortunately the extension has one big problem - the multisampled render to texture extension this uses will discard the frame buffer if other texture operations are used during rendering.

Is it possible to refactor `WebXRManager` so the old experiences continue working with and without `OCULUS_multiview`?

If we are not able to implement this without the developer having to know about it I don't think this PR will get merged.

mrdoob commented:
> Are you proposing that we detect problematic calls that cause flushes and then disable the extension (along with a warning)?

No no. I was hoping @snagy could clarify what exactly breaks when using `OCULUS_multiview` and what changes we need to do in the library and/or examples so using `OCULUS_multiview` doesn't break anything.

cabanier commented:
For good performance on mobile GPUs three should be doing these optimizations anyway; it's just more obvious if this extension is turned on (which is the reason they put this limitation in).
Ideally three would move to a multipass design where texture/mesh uploads happen first, followed by reflections and then the final render.

cabanier commented:
@snagy it looks like your patch still has conflicts

snagy commented:
> @snagy it looks like your patch still has conflicts

I'm happy you didn't come by when I accidentally pushed 366 pre-existing commits into this branch.  The force push was just patching up that mistake.

marcofugaro commented:
Hey, this optimization is really useful, would be cool to have it available.

HexaField commented:
This currently seems to not account for the skybox. It renders as a unit box centered on the camera, ie each face 0.5 units from the camera.

snagy commented:
> 

Just to be clear, by skybox do you mean the scene.background?

HexaField commented:
> > 
> 
> Just to be clear, by skybox do you mean the scene.background?

Sorry, yes.

snagy commented:
> This currently seems to not account for the skybox. It renders as a unit box centered on the camera, ie each face 0.5 units from the camera.

I finally got around to testing this and didn't see this problem - I made a scene with a cube map background and it rendered properly with and without multiview.  It looks like that code change a bit, (in pr https://github.com/mrdoob/three.js/pull/24805/ ) so it might have been fixed there.

HexaField commented:
> > This currently seems to not account for the skybox. It renders as a unit box centered on the camera, ie each face 0.5 units from the camera.
> 
> I finally got around to testing this and didn't see this problem - I made a scene with a cube map background and it rendered properly with and without multiview. It looks like that code change a bit, (in pr #24805 ) so it might have been fixed there.

After some more testing, we are only seeing this in the Meta Quest Pro. We have updated our fork of this PR, and the issue remains.

snagy commented:
> Update: We have modified L70 in WebGLBackground.js to have a size of 10000 for all parameters. This seems to be good enough for now.

I poked at this and thought about it a bit - I validated it was happening on my Quest Pro and then went back and saw the same behavior on my Q2.  It's incredibly uncomfortable, so I'm not sure if somehow I didn't notice it originally or if something else was happening.  I'll test this a few more times to see if there's more going on there.

Anyway, I think the behavior on the quest pro is 'correct' - it was creating a 1 meter box and then moving it to match the camera origin each frame.  Unfortunately for multiview, it's moving it to center on the spot between the cameras and then the multiview cameras still apply their view offsets when the box gets rendered, causing it to 'properly' offset per eye as though you've got a 1 meter box around your head.

The two solutions are either a) special case how the projection happens so the background projection isn't affected by camera position or b) just make the box really big.  (b) is the simpler option so I've updated this PR to include that.

myers commented:
I'm looking into how to speed up my WebXR app.  It's unclear to me why this was closed.  @mrdoob could you explain if there is a path forward for a PR like this one?
